src/utils/saml.ts

```diff
@@ -486,17 +753,75 @@ export const preValidateResponse: PreValidateResponseT = (
-      if (!ID || ID === "") {
-        throw new Error("Response must contain a non empty ID");
+      const IssueInstantValue = mainAttributeValidation(Response);
+      const InResponseTo = Response.getAttribute("InResponseTo");
+      if (!InResponseTo || InResponseTo === "") {
+        throw new Error("InResponseTo must contain a non empty string");
       }
-      const Version = Response.getAttribute("Version");
-      if (Version !== "2.0") {
-        throw new Error("Response version must be 2.0");
+      const Destination = Response.getAttribute("Destination");
+      if (!Destination || Destination === "") {
+        throw new Error("Response must contain a non empty Destination");
+      }
+      if (Destination !== samlConfig.callbackUrl) {
+        throw new Error(
+          "Destination must be equalt to AssertionConsumerServiceURL"
+        );
       }
+      const StatusElement = Response.getElementsByTagNameNS(
+        SAML_NAMESPACE.PROTOCOL,
+        "Status"
+      ).item(0);
+      if (!StatusElement || isEmptyNode(StatusElement)) {
+        throw new Error("Status element must be present");
+      }
+      const StatusCodeElement = StatusElement.getElementsByTagNameNS(
+        SAML_NAMESPACE.PROTOCOL,
+        "StatusCode"
+      ).item(0);
+      if (!StatusCodeElement) {
+        throw new Error("StatusCode element must be present");
+      }
+      const StatusCodeValue = StatusCodeElement.getAttribute("Value");
+      if (!StatusCodeValue || StatusCodeValue === "") {
+        throw new Error("StatusCode must contain a non empty Value");
+      }
+      if (StatusCodeValue !== "urn:oasis:names:tc:SAML:2.0:status:Success") {
+        throw new Error("Value attribute of StatusCode is invalid"); // TODO: Must be shown an error page to the user (26)
+      }
+      validateIssuer(Response);
+      const Assertion = Response.getElementsByTagNameNS(
+        SAML_NAMESPACE.ASSERTION,
+        "Assertion"
+      ).item(0);
+      if (!Assertion) {
+        throw new Error("Assertion element must be present");
+      }
+      const AssertionIssueInstantValue = mainAttributeValidation(Assertion);
+      assertionValidation(Assertion, samlConfig, InResponseTo);
+      samlConfig.cacheProvider?.get(InResponseTo, (err, value) => {
+        try {
+          if (err) {
+            throw new Error("Error reading the cache provider");
+          }
+          const RequestIssueInstant = new Date(value);
+          if (IssueInstantValue.getTime() < RequestIssueInstant.getTime()) {
+            throw new Error(
+              "Request IssueInstant must after Request IssueInstant"
+            );
+          }
+          if (
+            AssertionIssueInstantValue.getTime() < RequestIssueInstant.getTime()
+          ) {
+            throw new Error(
+              "Assertion IssueInstant must after Request IssueInstant"
+            );
+          }
+          callback(null, true);
+        } catch (e) {
+          callback(e);
```

Comment:
instead of throwing you can return an Either (or TaskEither if some check is async), wrap everything into a `tryCatch()` then `mapLeft(callback)`

Reply:
fixed